cpp/oneapi/dal/algo/kmeans/backend/gpu/kmeans_impl_dpc.cpp

```diff
@@ -0,0 +1,455 @@
+                                             cl::sycl::access::address_space::global_device_space>
+                        counter_atomic(counter_ptr[cl]);
+                    counter_atomic.fetch_add(1);
+                }
+            });
+    });
+    std::int32_t* value_ptr = empty_cluster_count.get_mutable_data();
+    auto empty_cluster_count_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on({ cluster_count_event });
+        cgh.parallel_for(
+            bk::make_multiple_nd_range_2d({ sg_size_to_set, 1 }, { sg_size_to_set, 1 }),
+            [=](sycl::nd_item<2> item) {
+                auto sg = item.get_sub_group();
+                const std::int64_t sg_id = sg.get_group_id()[0];
+                if (sg_id > 0)
+                    return;
+                const std::int64_t local_id = sg.get_local_id()[0];
+                const std::int64_t local_range = sg.get_local_range()[0];
+                std::int32_t sum = 0;
+                for (std::int64_t i = local_id; i < centroid_count; i += local_range) {
+                    sum += counter_ptr[i] == 0;
+                }
+                sum = reduce(sg, sum, sycl::ONEAPI::plus<std::int32_t>());
+                if (local_id == 0) {
+                    value_ptr[0] = sum;
+                }
+            });
+    });
+    return empty_cluster_count_event;
+}
+
+template <typename Float>
+sycl::event compute_objective_function(sycl::queue& queue,
+                                       const pr::ndview<Float, 2>& closest_distances,
+                                       pr::ndview<Float, 1>& objective_function,
+                                       const bk::event_vector& deps) {
+    ONEDAL_ASSERT(closest_distances.get_shape()[1] == 1);
+    ONEDAL_ASSERT(objective_function.get_shape()[0] == 1);
+    const Float* distance_ptr = closest_distances.get_data();
+    Float* value_ptr = objective_function.get_mutable_data();
+    const auto row_count = closest_distances.get_shape()[0];
+    const auto sg_size_to_set = get_gpu_sg_size(queue);
+    return queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(
+            bk::make_multiple_nd_range_2d({ sg_size_to_set, 1 }, { sg_size_to_set, 1 }),
+            [=](sycl::nd_item<2> item) {
+                auto sg = item.get_sub_group();
+                const std::int64_t sg_id = sg.get_group_id()[0];
+                if (sg_id > 0)
+                    return;
+                const std::int64_t local_id = sg.get_local_id()[0];
+                const std::int64_t local_range = sg.get_local_range()[0];
+                Float sum = 0;
+                for (std::int64_t i = local_id; i < row_count; i += local_range) {
+                    sum += distance_ptr[i];
+                }
+                sum = reduce(sg, sum, sycl::ONEAPI::plus<Float>());
+                if (local_id == 0) {
+                    value_ptr[0] = sum;
+                }
+            });
+    });
```

Comment:
Looks like it can be done via `reduction` primitive.

Reply:
it's a reduction of array to single value